File: packages/python/system.py
# encoding:utf-8
import psutil


def get_cpu_usage():
    return psutil.cpu_percent(interval=1)


def get_memory_usage():
    return psutil.virtual_memory().percent

# 获取该进程的cpu使用率
def get_process_cpu_usage():
    return psutil.Process().cpu_percent(interval=1)

# 聚合信息，返回一个字典
def get_system_info():
    return {
        "cpuUsage": get_cpu_usage(),
        "memoryUsage": get_memory_usage(),
        "processCpuUsage": get_process_cpu_usage()
    }

File: packages/server/src/utils/system.ts
const os = require('os')
const pidusage = require('pidusage')
const { exec } = require('child_process')
export const getMemoryUsage = () => {
  const totalMemory = os.totalmem()
  const freeMemory = os.freemem()
  const memoryUsage = (1 - freeMemory / totalMemory) * 100
  return memoryUsage
}

export const getProcessCpuUsage = (): Promise<number> => {
  return new Promise((resolve, reject) => {
    pidusage(process.pid, function (err, stats) {
      if (err) {
        console.error(err)
        reject(err)
        return
      }
      resolve(stats.cpu)
    })
  })
}

// 兼容windows和linux系统
export const getCPUUsage = () => {
  return new Promise((resolve, reject) => {
    exec(
      'typeperf -sc 1 "\\Processor(_Total)\\% Processor Time"',
      (error, stdout, stderr) => {
        if (error) {
          reject(error)
          return
        }
        // 解析输出，提取CPU使用率
        const outputLines = stdout.split('\n')
        const cpuUsageLine = outputLines[2]
        const cpuUsage = cpuUsageLine.split(',')[1].trim().replaceAll('"', '')
        resolve(Number(cpuUsage))
      },
    )
  })
}

export const getSystemInfo = async () => {
  const memoryUsage = getMemoryUsage()
  const cpuUsage = await getCPUUsage()
  const processCpuUsage = await getProcessCpuUsage()
  return {
    memoryUsage,
    cpuUsage,
    processCpuUsage,
  }
}
